File: app/src/primitives/Paint.ts
import { z } from "zod";

export const Color = z.object({
  value: z.array(z.number().min(0).max(255)).max(4),
});

const PaintStyleTypeOptions = ["Fill", "Stroke"] as const;

export const PaintStyleType = z.enum(PaintStyleTypeOptions);

const ColorWithDefault = Color.optional().default({ value: [0, 0, 0, 1] });

export const StrokeStyle = z.object({
  width: z.number().min(0).optional().default(10),
  color: ColorWithDefault,
  type: z.literal(PaintStyleType.Enum.Stroke),
});

export const FillStyle = z.object({
  color: ColorWithDefault,
  type: z.literal(PaintStyleType.Enum.Fill),
});

export const TextAlign = z.enum(["Left", "Center", "Right"]);

export const PaintStyle = z.discriminatedUnion("type", [
  StrokeStyle,
  FillStyle,
]);

export const Paint = z.object({
  style: PaintStyle,
});

export const TextPaint = z.object({
  style: PaintStyle,
  align: TextAlign,
  fontName: z.string().default("Helvetica-Bold"),
  size: z.number().min(0),
});

/* const NestedFillStyle = FillStyle.omit({ type: true }).default({});
const NestedStrokeStyle = StrokeStyle.omit({ type: true }).default({});

export const StrokeAndFillStyle = z.object({
  color: ColorWithDefault,
  type: z.literal(PaintStyleType.Enum.StrokeAndFill),
  fill: NestedFillStyle,
  stroke: NestedStrokeStyle,
}); */

File: app/src-tauri/src/animation/primitives/paint.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Color {
    value: (u8, u8, u8, f32),
}

impl Color {
    pub fn new(red: u8, green: u8, blue: u8, alpha: f32) -> Color {
        Color {
            value: (red, green, blue, alpha),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum PaintStyle {
    Fill(FillStyle),
    Stroke(StrokeStyle),
    StrokeAndFill(StrokeAndFillStyle),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Paint {
    pub style: PaintStyle,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TextPaint {
    pub style: PaintStyle,
    pub align: TextAlign,
    pub font_name: String,
    pub size: f32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct StrokeStyle {
    pub color: Color,
    pub width: f32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct StrokeAndFillStyle {
    pub stroke: StrokeStyle,
    pub fill: FillStyle,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FillStyle {
    pub color: Color,
}
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum TextAlign {
    Left,
    Center,
    Right,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FontDefinition {
    pub family_name: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Font {
    pub glyph_count: i32,
    pub weight: i32,
    pub style: String,
}
